### PR TITLE
osdep/io: add android related bullshit

### PR DESCRIPTION
**Pull request by jeeb**
This fixes >2GiB files with at least with API level 21.
See: https://android.googlesource.com/platform/bionic/+/master/docs/32-bit-abi.md
for the gritty details.

* Based on libavformat's things, except we do not care about API
  versions or NDKs without unistd.h, which contains all sorts of
  things that we utilize.
* Redefines lseek to always point to the 64bit version.
* Redefines fseeko to always point to lseek with fileno(f),
  as actual fseeko was only added with API level 24 (AKA Android 7).

_"Fuck Android"_

### CODE REVIEW

**Comment by jeeb** [comment]
The other alternative with regards to `fseeko` of course is that /stream/cache_file.c is made to use lseek instead of fseeko.

File: osdep/io.h [hunk 1]
```diff
@@ -34,14 +34,25 @@
...
 #  ifdef fseeko
 #   undef fseeko
 #  endif
-#  define fseeko(f,p,w) lseek(fileno(f),(p),(w))
+
+static inline int mp_fseeko(FILE* fp, off64_t offset, int whence) {
+    int ret = -1;
+    if ((ret = fflush(fp)) != 0) {
+        return ret;
+    }
```
Comment (wm4) [inline osdep/io.h]: Why is this needed?

**Comment by wm4** [comment]
Seems unavoidable. Not really happy with the inline function. I trust it'll go away in some years...

File: osdep/io.h (hunk 1 above)
Comment (jeeb) [reply osdep/io.h]: We're touching the file descriptor indecently from "underneath" the `FILE *` itself.

Discussion from last night with @rcombs:
```
01:36 < rcombs> JEEB: yeah that should work as long as nothing seeks,
                doesn't immediately seek back, and then attempts to read or
                write without first flushing
01:36 < rcombs> (i.e. as long as nothing produces a discrepancy between the
                FILE's buffer and the fd's position)
01:48 < rcombs> might want to fflush() to be sure
01:50 < rcombs> see what fseeko actually does:
https://github.com/android/platform_bionic/blob/38bcf2b704eb5038668759d2dd37115aea1dc437/libc/stdio/stdio.cpp#L496
01:51 < rcombs> (hint: it involves [basically] fflush())
01:52 < JEEB> so basically fflush and then seek?
01:52 < JEEB> or the other way+
01:53 < rcombs> flush then seek
01:53 < JEEB> k
01:56 < rcombs> also it won't work if anything uses ungetc, and doesn't
                clear the EOF flag
01:57 < rcombs> clearerr() will do that
01:57 < rcombs> (clear the EOF flag, that is)
```

**Comment by rcombs** [comment]
`lseek64` has always been in Bionic, so "at least with API level 21" is probably overly pessimistic

File: osdep/io.h (hunk 1 above)
Comment (wm4) [reply osdep/io.h]: Makes total sense. I assume there's no FILE API for this. Go on then.
